docs/proposals/synchronous-garbage-collection.md

```diff
@@ -0,0 +1,203 @@
+
+Synchronous Garbage Collection is a best-effort (see [unhandled cases](#unhandled-cases)) mechanism that allows user to determine if the garbage collection is done: after the API server receives a deletion request of an owning object, the object keeps existing in the key-value store until all its dependents are deleted from the key-value store by the garbage collector.
+
+Tracking issue: https://github.com/kubernetes/kubernetes/issues/29891
+
+# API Design
+
+## Standard Finalizers
+
+We will introduce a new standard finalizer:
+
+```go
+const GCFinalizer string = “DeletingDependents”
+```
+
+This finalizer indicates the object is terminating and is waiting for its dependents whose `OwnerReference.BlockOwnerDeletion` is true get deleted.
+
+## OwnerReference
+
+```go
+OwnerReference {
+     ...
+     // If true, AND if the owner has the "DeletingDependents" finalizer, then the owner cannot be deleted from the key-value store until this reference is removed.
+     // Defaults to false.
+     // To set this field, a user needs "delete" permission of the owner, otherwise 422 (Unprocessable Entity) will be returned.
+     BlockOwnerDeletion *bool
+}
+```
+
+The initial draft of the proposal did not include this field and it had a security loophole: a user who is only authorized to update one resource can set ownerReference to block the synchronous GC of other resources. Requiring users to explicitly set `BlockOwnerDeletion` allows the master to properly authorize the request.
+
+## DeleteOptions
+
+```go
+DeleteOptions {
+  …
+  // Whether and how garbage collection will be performed.
+  // Defaults to DefaultPropagationPolicy
+  PropagationPolicy *DeletePropagationPolicy
+}
+
+type DeletePropagationPolicy string
+
+const (
+    // Respects the existing garbage collection related finalizers on the object and the default garbage collection policy of the resource.
+    DefaultPropagationPolicy DeletePropagationPolicy = "Default"
+    // Orphans the dependents
+    OrphanDependents DeletePropagationPolicy = "Orphan"
+    // Deletes the object from the key-value store, the garbage collector will delete the dependents in the background.
+    DeleteDependentsInBackground DeletePropagationPolicy = "DeleteDependentsInBackground"
+    // The object exists in the key-value store until the garbage collector deletes all the dependents whose ownerReference.blockOwnerDeletion=true from the key-value store.
+    // API sever will put the "DeletingDependents" finalizer on the object, and sets its deletionTimestamp.
+    // This policy is cascading, i.e., the dependents will be deleted with GarbageCollectionSynchronous.
+    DeleteAfterBlockingDependentsAreDeleted DeletePropagationPolicy = "DeleteAfterBlockingDependentsAreDeleted"
```

Comment:
I think this name is not going to work. I prefer something short and not immediately obvious (like Foreground) so that users will read the longer, hopefully very clear, description in the documentation.